CON-84, CON-151: Add to Avail System Overview (#973)
maticnetwork/matic-docs: docs/avail/architecture/avail-system-overview.md

slug: avail-system-overview
---

## Modularity

Currently, monolithic blockchain architectures like that of Ethereum 
cannot efficiently handle execution, settlement, and data availability. 

Using modularity to scale blockchains is what the rollup-centric or 
application-specific chain model attempt to do. Sometimes, this works 
best when the settlement and data availability layer are on the same 

> layer, which is the approach Ethereum rollups take.

Reviewer: * Rollups modularize execution (execution is off-chain), but leave data availability in the monolithic chain (they submit transaction data to the L1). So they modularize _in part_.
* I don't think rollups "work best" necessarily - they make some tradeoffs, there are pros/cons. The rollup construction can be more secure (depending on the security of the data availability layer), but it is inherently more difficult to scale.

However, a granular design is designing different layers to be lightweight 
protocols, like microservices. Then, the overall network becomes a collection 
of loosely-coupled lightweight protocols. An example is a data availability 
layer that only specializes in data availability. Polygon Avail is a Substrate-based 

> application-specific chain for DA (data availability). 

Reviewer: Hmm I can see what you're trying to say, but I think it's confusing to describe Avail as an app chain, because app chains can be built _on top_ of Avail. So I think it would be better not to refer to Avail in that way.

:::note Substrate runtime

Modifications have been made to the actual Substrate runtime as opposed to 
the standard plug-and-play approach Substrate's FRAME offers. Please see the 

> [Avail architecture guide] for more details.

Reviewer: I'm not familiar with FRAME, but I would phrase this as: Although Avail is based on the Substrate codebase, it includes modifications to the block structure that prevent it from interoperating with other Substrate networks. Avail implements an independent network unrelated to Polkadot or Kusama.

:::

> Avail provides a high guarantee of data availability to any light client. 

Reviewer: To be more precise:
* Avail nodes do not make higher guarantees to light clients about DA than any other network.
* What Avail does provide is a way for light clients to "sample" small pieces of the block to verify it is actually there.
* Thus, the DA guarantee is something the client determines for itself—it doesn't have to trust nodes.

Also:
* As the number of light clients grows, they collectively sample the entire block (even though each client only samples a small %)
* Light clients actually form a P2P network amongst themselves
* Thus, after a block has been sampled, it becomes _highly available_—that is, even if the nodes were to go down (or attempt to censor a block), the light clients would be able to reconstruct the block by sharing the pieces amongst themselves.

Avail makes it possible to prove that block data is available without
downloading the whole block by leveraging Kate polynomial commitments,
erasure coding, and other technologies to allow light clients (which
download only the _headers_ of the chain) to efficiently and randomly
sample small amounts of the block data to verify its full
availability. However, there are fundamentally different primitives than 

> standard fraud-proof-based DA systems, which are explained [here].

Reviewer: Remove "standard" — fraud proofs are not really more standard than what we do.

### Enabling the next set of solutions

There are, of course, trade-offs that are made with different modularity 
approaches, but the overall goal is to maintain high security while being able 
to scale. Avail will take rollups to the next level as chains can allocate 
their data availability component to Avail. Avail also provides an alternative 
way to bootstrap [any] standalone chain, as chains can offload their data 

> availability. Transaction costs are also reduced; the primary cost 

Reviewer: ...as sovereign chains can be created that rely on Avail's validators to reach consensus on transaction data and ordering.

Reviewer: As for the tx costs point - it's true, but the way it is phrased is very much about validiums, whereas the point about bootstrapping sovereign rollups is about Avail-native chains. These are very different ways to use Avail, so I'd like to put more distance between them / make it clear these are different use cases.

Fundamentally, the reason Avail can remain cheap is that it is able to grow block size with a smaller impact to validator workload than a monolithic chain. When a monolithic chain increases block size, validators have to do a lot more work, because blocks have to be executed and state has to be calculated. Since avail has no execution environment at all, it's much cheaper to increase block size. (Not zero because we do need to calculate KZG commitments and generate proofs — but cheaper).

of committing a transaction through a rollup is the data availability cost for 
putting transaction data directly onto a layer one. Avail also makes sovereign 
rollups a possibility. Sovereign rollups on Avail allow for seamless upgrades, 
as updates can be pushed to application-specific nodes to upgrade the chain and, 
in turn, upgrade to new settlement logic. Whereas in a traditional environment, 

> direct updates would need to be made to contracts, or the network requires a fork.

Reviewer: I'd remove 'contracts' here - we're talking about fundamental network upgrades, not contracts.

:::info Avail does not have an execution environment

Avail does not run smart contracts, but it makes it possible
for other chains to make their transaction data available through Avail.
These chains may implement their execution environments of any kind,
EVM, Wasm, or anything else.

:::

:::info Avail doesn't care what the data is

Avail guarantees that block data is available but does not care about
what that data is. The data can be transactions but can take
on other forms too. Data availability on Avail is available 
for a window of time that it is required. For instance, beyond needing

> data or reconstruction, security is not compromised.

Reviewer: *for

Reviewer: This point about retention and the next paragraph about storage are good points, but shouldn't be mixed with the point about Avail being agnostic as to what the data is. Should be a separate point imo.

Storage systems, on the other hand, are designed to store data for long
periods, and include incentivization mechanisms to encourage users
to store data.

:::

## Validation

### Peer validation

Typically, Ethereum-like ecosystems contain three types of peers:

* **Validator nodes:**
  A validator collects transactions from the mempool, executes them, and generates a 
  candidate block that is appended to the network. The block contains a small block 
  header with the digest and metadata of the transactions in the block.
* **Full nodes:**
  The candidate block propogates to full nodes across the network for verification. 
  The nodes will re-execute the transactions contained in the candidate block. 
* **Light clients**
  Light clients only fetches the block header to use for verification, and will fetch 

>   transaction details from neighboring full nodes as needed.

Reviewer: Why not just talk about what the Avail nodes/clients are? Why talk about Ethereum?

For example - the Avail validators do not execute transactions. They simply package up the transactions (whatever they are) and construct candidate blocks, generating KZG commitments for the data. Other validators check that generated blocks are correct. Similarly, Avail full nodes do not execute anything.

While a secure approach, Avail addresses the limitations of this architecture to create 
robustness and increased guarantees. Light clients can be tricked into accepting blocks 
whose underlying data is not available. A block producer can include a malicious transaction 
in a block and not reveal its entire content to the network. As mentioned in the Avail
documentation, this is commonly known as the data availability problem.

Avail's network peers include:

* **Validator nodes:**
  Protocol incentivized full nodes that participate in the consensus.

* **Avail (DA) full nodes:**
   Nodes that download and make available all block data for all applications
   using Avail.

* **Avail (DA) light clients:**
   Clients that only download block headers, which randomly sample small parts of 
   the block to verify availability. They expose a local API to interact with the 
   Avail network.

:::info The goal of Avail is to not be reliant on full nodes to keep data available

>   The aim is to give the same DA guarantees to a light client as a full node. 

Reviewer: similar, not same

slug: avail-system-overview
---

## Modularity

Currently, monolithic blockchain architectures like that of Ethereum 

> cannot efficiently handle the execution, settlement, and data availability. 

Author: can leave "the" out

protocols, like microservices. Then, the overall network becomes a collection 
of loosely-coupled lightweight protocols. An example is a data availability 
layer that only specializes in data availability. Polygon Avail is a 
Substrate-based layer two blockchain for DA (data availability). 

:::note Substrate runtime

Although Avail is based on the Substrate codebase, it includes modifications 
to the block structure that prevent it from interoperating with other Substrate 
networks. Avail implements an independent network unrelated to Polkadot or Kusama.

:::

Avail provides a high guarantee of data availability to any light client,
but do not make higher guarantees to light clients about DA than any other 
network. Avail focuses on making it possible to prove that block data is 
available without downloading the whole block by leveraging Kate polynomial 
commitments, erasure coding, and other technologies to allow light clients (which
download only the _headers_ of the chain) to efficiently and randomly
sample small amounts of the block data to verify its full
availability. However, there are fundamentally different primitives than 

> fraud-proof-based DA systems, which are explained [here].

Author: ```suggestion
fraud-proof-based DA systems, which are explained [here](https://blog.polygon.technology/the-data-availability-problem-6b74b619ffcc/).
```

> fraud-proof-based DA systems, which are explained [here].

Author: updated

:::info Providing data availability

The DA guarantee is something a client determines for itself; it does not have to 
trust nodes. As the number of light clients grows, they collectively sample the entire 
block (even though each client only samples a small percentage). Light clients eventually 
form a p2p network amongst themselves; thus, after a block has been sampled, it becomes 
highly available—that is, even if the nodes were to go down (or attempt to censor a block), 
the light clients would be able to re-construct the block by sharing the pieces amongst 
themselves.

:::

### Enabling the next set of solutions

Avail will take rollups to the next level as chains can allocate 
their data availability component to Avail. Avail also provides an alternative 

> way to bootstrap [any] standalone chain, as chains can offload their data 

Author: ```suggestion
way to bootstrap any standalone chain, as chains can offload their data 
```

> way to bootstrap [any] standalone chain, as chains can offload their data 

Author: updated

increases block size, validators have to do a lot more work because blocks have to 
execute, and state has to be calculated. Since Avail has no execution environment, 
it is much cheaper to increase the block size. The cost is not zero because of the need to 
calculate KZG commitments and generate proofs, but still inexpensive. 

Avail also makes sovereign rollups a possibility. Users can create sovereign chains that 
rely on Avail's validators to reach consensus on transaction data and order. 
Sovereign rollups on Avail allow for seamless upgrades, as users can push updates to 
application-specific nodes to upgrade the chain and, in turn, upgrade to new settlement 
logic. Whereas in a traditional environment, the network requires a fork.

:::info Avail does not have an execution environment

Avail does not run smart contracts but allows other chains to complete their transaction 
data available through Avail. These chains may implement their execution environments of 
any kind, EVM, Wasm, or anything else.

:::

:::info Avail doesn't care what the data is for

Avail guarantees that block data is available but does not care about

> what that data is. The data can be transactions but can take

Author: we are being specific about transaction data

* **Light clients**
  Light clients only fetch the block header to use for verification and will fetch 
  transaction details from neighboring full nodes as needed.

While a secure approach, Avail addresses the limitations of this architecture to create 
robustness and increased guarantees. Light clients can be tricked into accepting blocks 
whose underlying data is unavailable. A block producer can include a malicious transaction 
in a block and not reveal its entire content to the network. As mentioned in the Avail
documentation, this is known as the data availability problem.

Avail's network peers include:

* **Validator nodes:**
  Protocol incentivized full nodes that participate in the consensus. Validator nodes
  on Avail do not execute transactions. They package up arbitrary transactions and 
  construct candidate blocks, generating KZG commitments for the data. 
  > Other validators check that generated blocks are correct.

* **Avail (DA) full nodes:**
   Nodes that download and make available all block data for all applications
   using Avail. Similarly, Avail full nodes do not execute transactions.

* **Avail (DA) light clients:**

>    Clients that only download block headers randomly sample small parts of 

Reviewer 2: ```suggestion
   Clients that only download block headers and random samples of 
```

Author: suggestion doesn't work with how random sampling is supposed to be mentioned

   the block to verify availability. They expose a local API to interact with the 
   Avail network.

:::info The goal of Avail is not to be reliant on full nodes to keep data available

  The aim is to give similar DA guarantees to a light client as a full node. 

>   Users are encouraged to use Avail light clients. However, users can run Avail 

Reviewer 2: ```suggestion
  Users are encouraged to use Avail light clients. However, they can still run Avail 
```

are, the greater the block size (and throughput) can securely exist. Meaning, non-consensus 
nodes can contribute to the throughput and security of the network.

### Transaction settlement

Avail will use a settlement layer built with Polygon Edge. The settlement layer 
provides an EVM-compatible blockchain for rollups to store their data and perform 
dispute resolution. The settlement layer utilizes Polygon Avail for its DA. 
When rollups are using a settlement layer, they also inherit all the 
DA properties of Avail.

:::note Different ways to settle

There are different ways to use Avail, and the validiums will not use the 
settlement layer, but rather settle on Ethereum.

:::

Avail offers data hosting and ordering. The execution layer will likely come from 
multiple off-chain scaling solutions or legacy execution layers. The settlement
layer takes on the verification and dispute resolution component.

## Resources

- [Introduction to Avail by Polygon blog post](https://medium.com/the-polygon-blog/introducing-avail-by-polygon-a-robust-general-purpose-scalable-data-availability-layer-98bc9814c048).

> - [Polygon Talks: Polygon Avail]

Reviewer 2: no link

Author: updated

> - [Polygon Talks: Polygon Avail]

Author: ```suggestion
- [Polygon Talks: Polygon Avail](https://www.youtube.com/watch?v=okqMT1v3xi0)
```

_attestation bridge_ that will post attestations of data available on
Avail to Ethereum, thus allowing validiums to be built.

## State verification

### Block verification -> DA verification

Instead of Avail validators verifying the application state, they concentrate 
on ensuring the availability of posted transaction data and provides transaction 

> ordering. A block is considered valid only if the data behind that block is available.

Reviewer: Not really wrong, but here's some additional context:

* Avail validators take incoming transactions, order them, construct a block, and propose it.
* The block contains special features especially for DA—erasure coding, and KZG commitments. And it's in a special format so that clients are able to do random sampling later, as well as being able to download only a single application's transactions.
* Anyway - they propose a block, and other validators verify the block is correct - well formed, KZG commitments check out, data is there, etc.

Requiring data to be available prevents block producers from releasing block headers
without releasing the data behind them, as this prevents clients from reading the 
transactions necessary to compute the state of their applications. Avail uses data 
availability verification to address this through DA checks which utilize erasure 

> codes; these checks are heavily used in data redundancy design.

Reviewer: * The DA verification is performed by clients via sampling. It's not something validators or full nodes do, exactly.
* Erasure codes effectively duplicate data in such a way that if part of a block is suppressed, clients can reconstruct that part by using another part of the block. This means that a node trying to suppress that part would actually need to suppress _a lot more_. The technique is used for example in CD-ROMs, and in multi-disk (RAID) arrays (so for example if a hard drive dies, it can be replaced and re-constructed from the data on other disks).

Reviewer: I mean—if data was indeed unavailable, then validators would not sign off on a block, that is true—but it's also true of any other chain. The thing that makes Avail unique isn't that _validators_ check the data, but rather that the chain is structured so that _anyone_ can check DA without actually downloading the data.

DA checks require each light client to sample a very small number of random chunks 
from each block in the chain. A set of light clients can collectively sample the 
entire blockchain in this manner. Consequently, the more non-consensus nodes there are 
the greater the block size (and throughput) can securely exist. Meaning, non-consensus 
nodes can contribute to the throughput and security of the network.

### Transaction settlement

Avail will use a settlement layer built with Polygon Edge. The settlement layer 
provides an EVM-compatible blockchain for rollups to store their data and perform 
dispute resolution. The settlement layer utilizes Polygon Avail for its own DA. 
When rollups are using a settlement layer, they also inherit all the 

> DA properties of Avail.

Reviewer: This is all correct - but note that there are different ways to use Avail, and the validium use case (referenced way above) will not use this settlement layer (validiums will settle on Ethereum). Just FYI!
